cpp/dolfinx/fem/dofmapbuilder.cpp

```diff
@@ -136,147 +136,223 @@ reorder_owned(const std::vector<dofmap_t>& dofmaps, std::int32_t owned_size,
 }
 
 //-----------------------------------------------------------------------------
-
 /// Build a simple dofmap from ElementDofmap based on mesh entity
 /// indices (local and global)
 ///
 /// @param [in] mesh The mesh to build the dofmap on
 /// @param [in] topology The mesh topology
 /// @param [in] element_dof_layout The layout of dofs on each cell type
-/// @return Returns: * dofmap for first element type [0] (local to the process)
+/// @return Returns: * dofmaps for each cell type (local to the process) as
+/// (width, [dofs])
 ///                  * local-to-global map for each local dof
 ///                  * local-to-entity map for each local dof
-/// Entities are represented as {dimension, mesh entity index}.
-std::tuple<dofmap_t, std::vector<std::int64_t>,
-           std::vector<std::pair<std::int8_t, std::int32_t>>>
-build_basic_dofmap(const mesh::Topology& topology,
-                   const fem::ElementDofLayout& element_dof_layout)
+///                  * index maps for each entity type in dofmaps
+///                  * global offset for dofs on this process
+/// Entities are represented as {index_map number, mesh entity index}.
+std::tuple<std::vector<dofmap_t>,
+           std::vector<std::int64_t>,
+           std::vector<std::pair<std::int8_t, std::int32_t>>,
+           std::vector<std::shared_ptr<const common::IndexMap>>, std::int64_t>
+build_basic_dofmaps(
+    const mesh::Topology& topology,
+    const std::vector<fem::ElementDofLayout>& element_dof_layouts)
 {
   // Start timer for dofmap initialization
   common::Timer t0("Init dofmap from element dofmap");
 
   // Topological dimension
   const std::size_t D = topology.dim();
+  const std::size_t num_cell_types = topology.entity_types(D).size();
+
+  LOG(INFO) << "Checking required entities per dimension";
+
+  // Find which dimensions (d) and entity types (et) are required
+  // and the number of dofs which are required for each (d, et) combination.
+  // Also store the IndexMaps and local offsets for each.
+  std::vector<std::pair<std::int8_t, std::int8_t>> required_dim_et;
```

Comment:
Way a `std::pair` when types are the same (`std::int8_t`)? 

`et` isn't informative.

Reply:
I guess I could use `std::array<std::int8_t, 2>` but I think `pair` shows the intent better, as the two values refer to different things.